### PR TITLE
Fix embedding_bag CPU scale_grad_by_freq using wrong index count

### LINKED ISSUE #190063: embedding_bag CPU dense backward scales gradients by the wrong frequency for mixed-count indices
### 🐛 Describe the bug

In `_embedding_bag_dense_backward_cpu_sum_mean`, the `scale_grad_by_freq`
scaling indexes the counts array with the unique-group loop counter instead of
the index value:

https://github.com/pytorch/pytorch/blob/26628087023e5102381849e9a27c3f09b6d485a4/aten/src/ATen/native/EmbeddingBag.cpp#L1572-L1574

```cpp
            if (scale_grad_by_freq) {
              scale /= counts[indices_data[i]];
```

`i` iterates over `next_unique_index_idx` (one step per unique index), while
the current element position is `j` and the current index value is `index`.
`indices_data[i]` therefore reads an unrelated position of the sorted indices
array. The result: a once-occurring index that sorts after a repeated one is
scaled by the repeated index's count.

```python
import torch

w = torch.ones(4, 3, requires_grad=True)
idx = torch.tensor([1, 1, 3])   # index 1 twice, index 3 once
offs = torch.tensor([0])
torch.nn.functional.embedding_bag(idx, w, offs, mode="sum",
                                  scale_grad_by_freq=True).sum().backward()
print(w.grad[:, 0].tolist())
```

Output on 2.13.0 and a source build of current main:

```text
[0.0, 1.0, 0.0, 0.5]
```

Row 1 is correct (`2 occurrences / count 2 = 1`), but row 3 should be `1.0`
(one occurrence, count 1) and is `0.5`: it was divided by `counts[indices_data[1]]
= counts[1] = 2`. The fix is `counts[index]`.

Found while implementing the corresponding MPS scaling (#190061 / #190062);
the MPS implementation divides by the true frequency, so the two backends
diverge on this input.

### Versions

PyTorch 2.14.0a0+git2662808 (source build of main) and 2.13.0; macOS 26.5.2 (arm64), Python 3.12.13. CPU-only bug, platform-independent.

cc @ezyang @albanD @gqchen @nikitaved @soulitzer @Varal7 @bobrenjc93 @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01

### PR DESCRIPTION
**Pull request by Ansh-Karnwal**
Fixes #190063.

### Root cause

In `_embedding_bag_dense_backward_cpu_sum_mean`, the `scale_grad_by_freq` divisor was `counts[indices_data[i]]`, where `i` is the outer loop counter that steps once per *unique* index (`next_unique_index_idx`), not the position of the current element. `indices_data[i]` therefore reads an unrelated slot of the sorted indices array rather than the current group's index value.

The current group's index value is already computed as `index` (`indices_data[indices_start]`), and `counts` is keyed by index value (`compute_counts` does `counts[indices_data[i]]++`). The correct divisor is `counts[index]`. This matches the non-bagged path in `Embedding.cpp` (`counts[k]`) and the CUDA/MPS backends, which were already correct; the CPU `embedding_bag` path was the lone outlier.

### Symptom

A once-occurring index that sorts after a repeated index gets divided by the repeated index's count. For indices `[1, 1, 3]` with `mode="sum"`, row 3 of the gradient was `0.5` (divided by count 2) instead of `1.0` (count 1).

This was discovered by @bogatyy while implementing the corresponding MPS scaling (#190062); the MPS backend divides by the true frequency, so CPU and MPS diverged on this input. This PR fixes the CPU side.

### Test plan

Reproducer from the issue now returns the correct gradient:

```python
import torch
w = torch.ones(4, 3, requires_grad=True)
torch.nn.functional.embedding_bag(torch.tensor([1,1,3]), w, torch.tensor([0]),
    mode='sum', scale_grad_by_freq=True).sum().backward()
print(w.grad[:,0].tolist())  # [0.0, 1.0, 0.0, 1.0]
```

New device-generic regression test (sum and mean modes, mixed-count indices):

```
python test/nn/test_embedding.py -k scale_grad_by_freq_mixed_counts -v
```

The test discriminates: rebuilt against the pre-fix code it fails (row 3 off by 0.5 on CPU) while CUDA passes; with the fix it passes on both. Broader sweep shows no regressions:

```
python test/nn/test_embedding.py -k embedding_bag -v   # 158 passed, 28 skipped
```
This PR description was formatted with AI assistance for readability.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01

### CODE REVIEW

**Comment by Skylion007** [comment]
@claude can you review this PR and double check the fix is correct?

**Review by Skylion007 — APPROVED** [review]

File: test/nn/test_embedding.py [hunk 1]
```diff
@@ -986,6 +986,29 @@ def grad_at_r(idx_dtype):
...
+    @dtypes(torch.float32, torch.float64)
+    def test_embedding_bag_scale_grad_by_freq_mixed_counts(self, device, dtype):
+        # scale_grad_by_freq must divide each index's gradient by that index's
+        # own occurrence count. Use mixed counts (index 1 twice, index 3 once)
+        # so a once-occurring index sorts after a repeated one: this is the case
+        # where indexing the counts array by loop position rather than by index
+        # value produces the wrong scale.
+        # sum: grad row = occurrences / count. mean: additionally / bag_size (3).
+        expected_grads = {"sum": (1.0, 1.0), "mean": (1.0 / 3, 1.0 / 3)}
+        for mode, (row1, row3) in expected_grads.items():
```
Comment (Skylion007) [inline test/nn/test_embedding.py:1000]: Really minor nit but wrapping this with self.subTest would be great

Comment (Skylion007) [reply test/nn/test_embedding.py:1000]: Or better yet parameterizing by it
